### DISCUSSION TITLE
what should I expect with server routes having parameters and a catchall?

### DISCUSSION THREAD

**Discussion by sig9** [nitrojs/nitro#324] [Q&A] [answered]
I have the following layout in my /server/api directory

![image](https://user-images.githubusercontent.com/2271605/176963297-d68e2712-e33d-4bd8-8af9-626ceadeef16.png)

I was hoping that if someone hit the /api/db/collection/ url that the [...] catchall would be hit.

Instead what I'm seeing is that [id].get.ts is called with id set to an empty string. 

Not sure if there is a better way or if this might be an issue.

**Comment by sig9** [comment, marked as answer]
Right after posting I thought to try an alternative. 

![image](https://user-images.githubusercontent.com/2271605/176964257-4b62a462-a77b-490e-b797-86cb0d4369be.png)

This works great if I GET /api/db/collection (without the trailing slash) In this case db/[collection].get.ts is called.

If I have a trailing slash... /api/db/collection/ invokes db/[collection]/[id].get.ts with an empty string for id

I think ? I"m doing it right, but not 100% sure. Any suggestions appreciated. I'm going to mark this as answered.

**Reply by OptimusePrime** [reply]
If you want to call an API route such as `/api/user` and put the catch-all in the `user` folder, you _must_ put the slash at the end (`/api/user/`). Otherwise, it will throw a `404 Not Found` error. It's the opposite with the `index.get.ts`. If you don't put a slash (`/api/user`) it'll work, but if you do (`/api/user/`), it won't. 

You can use either or both, but not using a slash at the end seems more natural. So, you can put everything concerning the `collection` inside one folder. Is the `collection` named `[collection]`, or are those brackets created by the editor? If it has square brackets in the name, that might be the potential source of the error.

**EDIT:** I just tried it again, and it won't throw a `404 Not Found` if you didn't put a slash at the end if you have a catch-all further up the routing chain. For example, if you had a catch-all inside the `user` folder and a catch-all inside the `api` folder, and used the `/api/user` route, it'd default to the catch-all inside the `api` (It defaults to the catch-all (if one exists) inside the parent folder.); it'd be as though you accessed `/api/` instead of `/api/user/`. If the catch-all is inside the root `routes` folder, it'll work with and without a slash at the end.

**Reply by sig9** [reply]
thanks @OptimusePrime 
I have not tried index.get.ts as a catchall. Thanks for the suggestion.

**Reply by sig9** [reply]
The [collection] syntax is a route parameter helping me create a generic API for multiple collections. The following url's would match
/db/examples/
/db/users/

in the setup I access the collection parameter as

``` json
export default defineEventHandler(async (event) => {
    const container = event.context.params.collection
    ... database access code here.
```
